File: src/lessons/welcome/ArrayEntity.java
package lessons.welcome;

import java.awt.Color;

public class ArrayEntity extends jlm.universe.bugglequest.SimpleBuggle {
	@Override
	public void setX(int i)  {
		if (isInited())
			throw new RuntimeException("setX(int) forbidden in this exercise");
	}
	@Override
	public void setY(int i)  { 
		if (isInited())
			throw new RuntimeException("setY(int) forbidden in this exercise");
	}
	@Override
	public void setPos(int i,int j)  { 
		if (isInited())
			throw new RuntimeException("setPos(int,int) forbidden in this exercise");
	}

	/* BEGIN SOLUTION */
	void mark(Color c){
		setBrushColor(c);
		brushDown();
		brushUp();
	}
	
	public void run() {
		Color[] colors = new Color[getWorldHeight()];
		
		/* read the colors */
		for (int i=0;i<getWorldHeight();i++) {
			colors[i]=getGroundColor();
			forward();
		}
		
		/* duplicate the pattern */
		for (int i=1; i<getWorldWidth();i++) {
			turnLeft();
			forward();
			turnRight();
			forward();
			makeLine(colors);
		}
	}
	
	void makeLine(Color[] colors) {
		for (int i=0;i<getWorldWidth();i++) {
			mark(colors[i]);
			forward();
		}
	}
	/* END TEMPLATE */
}

File: src/lessons/welcome/array/basics/ArrayEntity.py
def setX(i):
	errorMsg("Sorry Dave, I cannot let you run setX(i) in this exercise")

def setY(i):
	errorMsg("Sorry Dave, I cannot let you run setY(i) in this exercise")
def setPos(x,y):
	errorMsg("Sorry Dave, I cannot let you run setPos(x,y) in this exercise")

# BEGIN TEMPLATE
# BEGIN SOLUTION
def mark(color):
	setBrushColor(color)
	brushDown()
	brushUp()

colors = []

# read the colors
for i in range(getWorldHeight()):
	colors.append(getGroundColor())
	forward()

def makeLine (colors):
	for i in range(getWorldWidth()):
		mark(colors[i])
		forward()

# duplicate the pattern
for i in range(getWorldWidth()-1):
	turnLeft()
	forward()
	turnRight()
	forward()
	makeLine(colors)

# END SOLUTION
# END TEMPLATE
